### PR TITLE
Complete set_attn_processor for prior and vae

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This adds cleans up the tests a bit and makes sure that:
- `UNet2DConditionModel`
- `PriorTransformer`
- `AutoencoderKL`

all have the same features. This PR also cleans up the tests a bit and adds a whole test suite for the prior transformer.

### CODE REVIEW

File: tests/models/test_models_unet_1d.py [hunk 1]
```diff
@@ -153,7 +154,7 @@ def test_unet_1d_maestro(self):
         assert (output_max - 0.0607).abs() < 4e-4
 
 
-class UNetRLModelTests(ModelTesterMixin, unittest.TestCase):
+class UNetRLModelTests(ModelTesterMixin, UNetTesterMixin, unittest.TestCase):
     model_class = UNet1DModel
```
Comment (patrickvonplaten) [inline tests/models/test_models_unet_1d.py:158]:
```suggestion
    model_class = UNet1DModel
    main_input_name = "sample"
```

File: tests/models/test_models_unet_2d.py [hunk 2]
```diff
@@ -68,7 +69,7 @@ def prepare_init_args_and_inputs_for_common(self):
         return init_dict, inputs_dict
 
 
-class UNetLDMModelTests(ModelTesterMixin, unittest.TestCase):
+class UNetLDMModelTests(ModelTesterMixin, UNetTesterMixin, unittest.TestCase):
     model_class = UNet2DModel
```
Comment (patrickvonplaten) [inline tests/models/test_models_unet_2d.py:73]:
```suggestion
    model_class = UNet2DModel
    main_input_name = "sample"
```

File: tests/models/test_models_unet_2d.py [hunk 3]
```diff
@@ -182,7 +183,7 @@ def test_output_pretrained(self):
         self.assertTrue(torch_all_close(output_slice, expected_output_slice, rtol=1e-3))
 
 
-class NCSNppModelTests(ModelTesterMixin, unittest.TestCase):
+class NCSNppModelTests(ModelTesterMixin, UNetTesterMixin, unittest.TestCase):
     model_class = UNet2DModel
```
Comment (patrickvonplaten) [inline tests/models/test_models_unet_2d.py:187]:
```suggestion
    model_class = UNet2DModel
    main_input_name = "sample"
```

**Review by kashif — APPROVED** [review]
i love the tests!

File: src/diffusers/models/autoencoder_kl.py [hunk 4]
```diff
@@ -12,13 +12,14 @@
...
 from dataclasses import dataclass
-from typing import Optional, Tuple, Union
+from typing import Dict, Optional, Tuple, Union
 
 import torch
 import torch.nn as nn
 
 from ..configuration_utils import ConfigMixin, register_to_config
 from ..utils import BaseOutput, apply_forward_hook
+from .attention_processor import AttentionProcessor, AttnProcessor
```
Comment (sayakpaul) [inline src/diffusers/models/autoencoder_kl.py:22]: Should we maybe also implement this automatic discoverability of `AttnProcessor2_0`? 

https://github.com/huggingface/diffusers/blob/231bdf2e560a57976749b55a941bcf3f311df6fc/src/diffusers/models/attention_processor.py#L160

File: src/diffusers/models/prior_transformer.py [hunk 5]
```diff
@@ -8,6 +8,7 @@
 from ..configuration_utils import ConfigMixin, register_to_config
 from ..utils import BaseOutput
 from .attention import BasicTransformerBlock
+from .attention_processor import AttentionProcessor, AttnProcessor
```
Comment (sayakpaul) [inline src/diffusers/models/prior_transformer.py:11]: Same as https://github.com/huggingface/diffusers/pull/3796/files#r1231049627

File: tests/models/test_models_unet_2d.py [hunk 6]
```diff
@@ -182,8 +184,9 @@ def test_output_pretrained(self):
         self.assertTrue(torch_all_close(output_slice, expected_output_slice, rtol=1e-3))
 
 
-class NCSNppModelTests(ModelTesterMixin, unittest.TestCase):
+class NCSNppModelTests(ModelTesterMixin, UNetTesterMixin, unittest.TestCase):
```
Comment (sayakpaul) [inline tests/models/test_models_unet_2d.py:187]: Curiosity: 

What does it mean by `NCSNpp`?

File: tests/models/test_models_prior.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        # Since the VAE Gaussian prior's generator is seeded on the appropriate device,
+        # the expected output slices are not the same for CPU and GPU.
+        expected_output_slice = torch.tensor([-1.3436, -0.2870, 0.7538, 0.4368, -0.0239])
+        self.assertTrue(torch_all_close(output_slice, expected_output_slice, rtol=1e-2))
+
+
+@slow
+class PriorTransformerIntegrationTests(unittest.TestCase):
+    def get_file_format(self, seed, shape):
+        return f"gaussian_noise_s={seed}_shape={'_'.join([str(s) for s in shape])}.npy"
```
Comment (sayakpaul) [inline tests/models/test_models_prior.py:140]: Clever code hahaha.

**Review by sayakpaul — APPROVED** [review]
Exceptionally clean!

File: tests/models/test_models_unet_2d.py (hunk 6 above)
Comment (patrickvonplaten) [reply tests/models/test_models_unet_2d.py:187]: https://github.com/yang-song/score_sde_pytorch

File: tests/models/test_models_unet_1d.py [hunk 8]
```diff
@@ -153,8 +154,9 @@ def test_unet_1d_maestro(self):
         assert (output_max - 0.0607).abs() < 4e-4
 
 
-class UNetRLModelTests(ModelTesterMixin, unittest.TestCase):
+class UNetRLModelTests(ModelTesterMixin, UNetTesterMixin, unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply tests/models/test_models_unet_1d.py:157]: Tests a RL model we've integrated

File: tests/models/test_models_prior.py (hunk 7 above)
Comment (patrickvonplaten) [inline tests/models/test_models_prior.py:140]:
```suggestion
```

File: src/diffusers/models/autoencoder_kl.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/diffusers/models/autoencoder_kl.py:22]: I don't fully follow here. Note that by default the ATTN 2.0 is always chosen whenever someone uses the `Attention` class which is done for both the AutoEncoder and the Prior, see:
- https://github.com/huggingface/diffusers/blob/77f9137f103b04f137b6c487814ea58599ebc200/src/diffusers/models/attention_processor.py#L158
and:
- https://github.com/huggingface/diffusers/blob/231bdf2e560a57976749b55a941bcf3f311df6fc/src/diffusers/models/attention.py#L85

**Pull request merged**